instrumentation/opentelemetry-instrumentation-urllib/src/opentelemetry/instrumentation/urllib/__init__.py

```diff
@@ -154,7 +154,7 @@ def _instrumented_open_call(
             inject(type(headers).__setitem__, headers)
 
             token = context.attach(
-                context.set_value(_SUPPRESS_URLLIB_INSTRUMENTATION_KEY, True)
+                context.set_value(_SUPPRESS_HTTP_INSTRUMENTATION_KEY, True)
```

Comment:
Does this mean if urllib instrumentation is used then http.client instrumentation (if present) will be suppressed? I don't think any instrumentation should dictate if another instrumentation should be silenced or not. What if a user actually wants both? May be there are issues with http.client or with urllib and users would want to detect exactly which or may be http.client has adds lower level information like DNS, connection, SSL-handshake, etc related info. Should urllib really automatically suppress it? 

I think it should be left to the user whether to any specific instrumentation should be enabled or not.

Reply:
Hm.., i guess it depends on what one actually expects to see, the span produced by the instrumentation or the outgoing HTTP call. 
The span produced by an instrumentations like e.g. for HTTP clients is usually the last in process span before the trace is continued in some other process (on probably some other host). Hence, the `inject` call to propagate the context to the next process. Without supressing the lower level HTTP instrumentations, each would separatly inject their span into the HTTP headers. From a functionality point of view this is probably not a problem since traceparent and tracecontext are going to ge overwritten (as far is a saw) but will introduce additional performance overhead. Not sure if a hierarchy of subsequent spans of kind `CLIENT` would cause problems though.

Looking at it from a pure span point of view it raises the question i guess why some instrumentations suppress library internal calls in the first place, e.g. the call from `Session.request` to `Session.send` in the `requests` library (and i think urllib instrumentor is doing something similar). The produced spans are more or less the same but would still provide additional info through timing and span hierarchy. For a user of the library this is probably of less to no interest but i guess the same can be said about spans produced by the libraries which are internally used by the instrumented library. E.g. `requests` and `urllib3` produce the same span with just the `InstrumentationInfo` being different.

If a user is only interested in the span for the outgoing HTTP (without the spans produced by the internal libraries) i think it is not always possible to achieve this by deciding if an instrumentation is enabled or not. E.g. in an application which uses a mix of `requests` and `urllib3` for communication. With an adapter `requests` can be configured to use a protocol other than HTTP for certain sessions. If one now wants to track the HTTP calls made with `requests`, `urllib3` plus the calls with `requests` using other protocol, HTTP calls made with `requests` would always generate two outgoing HTTP spans (requests + urllib3) without the possibility to deactivate this behavior.

Imo from a users point of view the outgoing HTTP call might of more interest than the internal workings of the `repquests`, `urllib`, ... libraries but i can change the PR to not suppress other instrumentations if you want .

Reply:
Internal library spans are a completely difference use case. If they are not silenced then every span exported/generated will generate more spans which will in-turn generate more spans. So even generating a single span will start a loop of spans being generated. So we *have to* silence the internal library spans.

However, I'm convinced by the argument that not silencing will result in a parent > child pair of spans that will both be of kind CLIENT. I still don't think the solution should be for one instrumentation to silence other instrumentations but I don't see a simpler solution to solve this without adding some advanced features.

So it's fine if urllib3 or requests silences http instrumentation in this PR. I'll create an issue to discuss this further and possibly improve it in future.

Reply:
OK we can probably solve in a simple enough way. 

What if urllib3 and requests never injected tracing context headers and always set span kind to `INTERNAL` and then added http instrumentation as a dependency. 

So whenever either urllib3 or requests is installed, http.client instrumentation is installed as well. If requests instrumentation is smart enough to never set kind to CLIENT and never inject tracing context when HTTP transport is used, this should cleanly solve the entire problem. 

In this case, we'd replace the current suppression calls with a comment that says if we ever build a http.client instrumentation, we should update urllib3 and requests to not generate client spans. 

Only issue I see with this is that these libs will hard depend on our http.client instrumentation so it'll be a bit weird (install other package, set env var to disable our http.client instrument) for users to swap out the http.client instrumentation with their own but I think this would be a very rare edge-case with a very viable solution.

What do you think?

Reply:
Opened an issue to discuss this further: https://github.com/open-telemetry/opentelemetry-python-contrib/issues/369

I'm OK with this PR to suppress http.client instrumentation in the meanwhile.

Reply:
@owais 
You have a changes requested for this PR.